lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/avro/IMessageDeserializer.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.wake.avro;
+
+import org.apache.avro.io.BinaryDecoder;
+import org.apache.reef.wake.MultiObserver;
+
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+
+/**
+ * Base interface for Avro message deserializer objects.
+ */
+public interface IMessageDeserializer {
```

Comment:
yes, we don't use avro in protocols yet. We use protobuf, which has much the same limitations as avro.

That said, I don't understand where the need to work with arbitrary, previously unknown avro types is entering the picture. Can't we enumerate all message types that go from Java to .NET and vice versa? 

If so, we can do what we've done in the Evaluator-Driver communications: Make one envelope message which has optional fields for each of those message types and an `enum` field to indicate what kind of message it is.

Side note: We do not, in fact have the `enum` in `EvaluatorHeartbeatProto` and `EvaluatorControlProto` which I now consider a mistake, as it needlessly complicates the message dispatch in Driver and Evaluator.

The upside of that approach is that the grand total of the Driver-Driver protocol is statically known and documented in that schema, which makes adding additional Driver implementations easy. The downside is that we have to rev the protocol for every new, removed or changed message.

If we in fact cannot enumerate all message types that need interchange at compile time, we have to follow the approach here. However, that would make adding new Drivers (e.g. in Python or R) a lot more challenging as they'd also have to support the "any message at any time" feature.